packages/@aws-cdk/aws-ec2/lib/client-vpn.ts

```diff
@@ -0,0 +1,409 @@
+  readonly transportProtocol?: ClientVpnEndpointProtocol;
+}
+
+export interface IClientVpnEndpoint extends cdk.IResource {
+  /**
+   * The id of the Client VPN Endpoint
+   */
+  readonly clientVpnEndpointId: string;
+  readonly routes: ClientVpnRoute[];
+  readonly targetNetworkAssociations: ClientVpnRoute[];
+  readonly authorizationRules: ClientVpnRoute[];
+}
+
+/**
+ * Define a Client VPN Endpoint
+ *
+ * A Client VPN endpoint is the resource you create and configure to enable and manage client VPN sessions.
+ * It is the destination endpoint at which all client VPN sessions are terminated.
+ *
+ * @resource AWS::EC2::ClientVpnEndpoint
+ */
+export class ClientVpnEndpoint extends cdk.Resource implements IClientVpnEndpoint {
+
+  public readonly clientVpnEndpointId: string;
+  public readonly routes: ClientVpnRoute[] = [];
+  public readonly targetNetworkAssociations: ClientVpnRoute[] = [];
+  public readonly authorizationRules: ClientVpnRoute[] = [];
+
+  constructor(scope: cdk.Construct, id: string, props: ClientVpnEndpointProps) {
+    super(scope, id);
+
+    if (!cdk.Token.isUnresolved(props.clientCidrBlock)) {
+      const cidrMatch = props.clientCidrBlock.match(CIDR_VALIDATION_REGEXES.ipv4);
+
+      if (!cidrMatch) {
+        throw new Error(`Invalid IPv4 CIDR: "${props.clientCidrBlock}"`);
+      }
+
+      if (!cidrMatch[2] || Number(cidrMatch[2]) < 22) {
+        throw new Error(`CIDR mask should be 22 or greater (got "${cidrMatch[2]}")`);
+      }
+    }
+
+    if (props.dnsServers && !cdk.Token.isUnresolved(props.dnsServers) && props.dnsServers.length > 2) {
+      throw new Error(`A Client VPN endpoint cannot have more than 2 DNS servers, got ${props.dnsServers.length}`);
+    }
+
+    const clientVpnEndpoint = new CfnClientVpnEndpoint(this, 'Resource', {
+      authenticationOptions: props.authenticationOptions.map(({options}) => options),
+      connectionLogOptions: props.connectionLog && props.connectionLog.options,
+      serverCertificateArn: props.serverCertificateArn,
+      clientCidrBlock: props.clientCidrBlock,
+      description: props.description,
+      splitTunnel: props.splitTunnel,
+      tagSpecifications: props.tagsSpecifications,
+      dnsServers: props.dnsServers,
+      transportProtocol: props.transportProtocol,
+    });
+
+    this.clientVpnEndpointId = clientVpnEndpoint.ref;
+  }
+
+  public addRoute(options: ClientVpnRouteOptions): ClientVpnRoute {
+    const route = new ClientVpnRoute(this, `Route-${this.routes.length}`, {
+      clientVpnEndpoint: this,
+      ...options,
+    });
+
+    this.routes.push(route);
+    return route;
+  }
+
+  public addTargetNetworkAssociation(subnet: ISubnet): ClientVpnTargetNetworkAssociation {
+    const association = new ClientVpnTargetNetworkAssociation(
+      this,
+      `TargetNetworkAssociation-${this.targetNetworkAssociations.length}`, {
+        clientVpnEndpoint: this,
+        subnet,
+      });
+
+    this.targetNetworkAssociations.push(association);
```

Comment:
This is being pushed both here and in the constructed class.

I'm not sure that I'm a fan. What's the use case for maintaining and publishing these objects through the public array?

Reply:
I was keeping track of them for the `this.targetNetworkAssociations.length`, but it doesn't feel great I agree. I didn't know what else to hook onto for the id generation.

Reply:
You can just keep the count, instead of the objects themselves.